src/Compilers/CSharp/Test/Semantic/Semantics/ExtensionTypeTests.cs

```diff
@@ -0,0 +1,4991 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics;
+
+public class ExtensionTypeTests : CompilingTestBase
+{
+    private static void VerifyNotExtension<T>(TypeSymbol type) where T : TypeSymbol
+    {
+        Assert.True(type is T, $"Found type '{type.GetType()}'");
+        Assert.False(type.IsExtension);
+        Assert.Null(type.ExtensionUnderlyingTypeNoUseSiteDiagnostics);
+        Assert.Empty(type.BaseExtensionsNoUseSiteDiagnostics);
+    }
+
+    // Verify things that are common for all extension types
+    private static void VerifyExtension<T>(TypeSymbol type, SpecialType specialType = SpecialType.None) where T : TypeSymbol
+    {
+        Assert.True(type is T);
+        Assert.True(type.IsExtension);
+        Assert.Null(type.BaseTypeNoUseSiteDiagnostics);
+        Assert.False(type.IsSealed); // PROTOTYPE
+        Assert.False(type.IsRecord);
+        Assert.False(type.IsRecordStruct);
+        Assert.False(type.IsReferenceType);
+        Assert.False(type.IsValueType);
+        Assert.False(type.IsTypeParameter());
+        Assert.False(type.IsAnonymousType);
+        Assert.False(type.IsEnumType());
+        Assert.False(type.IsErrorType());
+        Assert.Equal(specialType, type.SpecialType);
+        Assert.False(type.IsObjectType());
+        Assert.False(type.IsTupleType);
+        Assert.True(type.TupleElements.IsDefault);
+        Assert.Empty(type.InterfacesNoUseSiteDiagnostics());
+        Assert.Empty(type.AllInterfacesNoUseSiteDiagnostics); // PROTOTYPE
+        Assert.False(type.IsReadOnly);
+        Assert.False(type.IsRefLikeType);
+        Assert.False(type.IsUnsafe());
+        Assert.Equal(TypeKind.Extension, type.TypeKind);
+        Assert.False(type.IsInterfaceType());
+        Assert.False(type.IsAbstract);
+
+        if (type.ExtensionUnderlyingTypeNoUseSiteDiagnostics is { } underlyingType)
+        {
+            VerifyNotExtension<TypeSymbol>(underlyingType);
```

Comment:
Not sure if this will always be true for error scenarios? Seems like this information should be exposed in such locations.

Reply:
Currently, we only record the underlying type if it's valid.
This is similar to the behavior for interfaces (see `MakeOneDeclaredBases`) where invalid types (like a struct) won't get added to the interface list.
I'm hoping this should make consumption easier (no need to recheck every type someone uses this API).
That said, I added a PROTOTYPE comment in test to revisit/confirm in context of public APIs.